google/cloud/spanner/transaction.h

```diff
@@ -0,0 +1,156 @@
+ * write operations, tying them together, but is otherwise opaque to the user.
+ *
+ * All reads/writes in the transaction must be executed within the same
+ * session, and that session may have only one transaction active at a time.
+ *
+ * Spanner supports these transaction modes:
+ *   - ReadOnly. Provides guaranteed consistency across several reads, but does
+ *     not allow writes. Can be configured to read at timestamps in the past.
+ *     Does not need to be committed and does not take locks.
+ *   - ReadWrite. Supports reading and writing data at a single point in time.
+ *     Uses pessimistic locking and, if necessary, two-phase commit. May abort,
+ *     requiring the application to retry.
+ */
+class Transaction {
+ public:
+  /**
+   * Options for ReadOnly transactions.
+   */
+  class ReadOnlyOptions {
+   public:
+    // Strong: Guarantees visibility of the effects of all transactions that
+    // committed before the start of the reads.
+    ReadOnlyOptions();
+
+    // Exact Staleness: Executes all reads at `read_timestamp`.
+    ReadOnlyOptions(Timestamp read_timestamp);
+
+    // Exact Staleness: Executes all reads at a timestamp `exact_staleness`
+    // old. The actual timestamp is chosen soon after the reads are started.
+    ReadOnlyOptions(std::chrono::nanoseconds exact_staleness);
+
+   private:
+    friend Transaction;
+    google::spanner::v1::TransactionOptions_ReadOnly ro_opts_;
+  };
+
+  /**
+   * Options for ReadWrite transactions.
+   */
+  class ReadWriteOptions {
+   public:
+    // There are currently no read-write options.
+    ReadWriteOptions();
+
+   private:
+    friend Transaction;
+    google::spanner::v1::TransactionOptions_ReadWrite rw_opts_;
+  };
+
+  /**
+   * Options for "single-use", ReadOnly transactions, where Spanner chooses
+   * the read timestamp, subject to user-provided bounds. This allows reading
+   * without blocking.
+   *
+   * Because selection of the timestamp requires knowledge of which rows will
+   * be read, a single-use transaction can only be used with one read. See
+   * Client::Read() and Client::ExecuteSql(). SingleUseOptions cannot be used
+   * to construct an application-level Transaction.
+   */
+  class SingleUseOptions {
+   public:
+    // Strong or Exact Staleness: See ReadOnlyOptions.
+    SingleUseOptions(ReadOnlyOptions opts);  // implicit conversion
+
+    // Bounded Staleness: Executes all reads at a timestamp that is not
+    // before `min_read_timestamp`.
+    SingleUseOptions(Timestamp min_read_timestamp);
+
+    // Bounded Staleness: Executes all reads at a timestamp that is not
+    // before `NOW - max_staleness`.
+    SingleUseOptions(std::chrono::nanoseconds max_staleness);
+
+   private:
+    friend Transaction;
+    google::spanner::v1::TransactionOptions_ReadOnly ro_opts_;
+  };
+
+  /// @name Construction of read-only and read-write transactions.
+  ///@{
+  Transaction(ReadOnlyOptions opts);
```

Comment:
How do these constructors connect a transaction to a session and client?  Or do they create a transaction selector, but the transaction is not really created yet?

Reply:
Right ... the "transaction-selector to transaction-id" transformation is not made until the `Transaction` is used in the context of a particular client/session.  Then it is up to the client to only use the `Transaction` in that context.

Reply:
Ack.